### PR TITLE
Query entity

### PR DESCRIPTION
**Pull request by 07Himank**
### Describe your changes :

I worked on the ..... because ... created query as entity.

#
### Type of change :

- [x] Bug fix

Fix : #9660

### CODE REVIEW

File: openmetadata-spec/src/main/resources/json/schema/entity/data/query.json [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    "vote": {
+      "description": "Users can vote up to rank the popular queries.",
+      "type": "number",
+      "default": 1
+    },
+    "checksum": {
+      "description": "Checksum to avoid registering duplicate queries.",
+      "type": "string"
+    },
+    "queryDate": {
```
Comment (harshach) [inline openmetadata-spec/src/main/resources/json/schema/entity/data/query.json:39]: lets make this as a timestamp

File: openmetadata-spec/src/main/resources/json/schema/entity/data/query.json [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  "type": "object",
+  "javaType": "org.openmetadata.schema.entity.data.Query",
+  "javaInterfaces": ["org.openmetadata.schema.EntityInterface"],
+  "properties": {
+    "id": {
+      "description": "Unique identifier of the query.",
+      "$ref": "../../type/basic.json#/definitions/uuid"
+    },
+    "query": {
+      "description": "query.",
```
Comment (harshach) [inline openmetadata-spec/src/main/resources/json/schema/entity/data/query.json:15]: description should be "SQL Query definition"

**Comment by harshach** [comment]
Thanks @07Himank for the PR. We should have migration script that can load existing queries data into the new table and delete the old data.

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java [hunk 3]
```diff
@@ -1899,6 +1925,47 @@ List<String> listPrefixesAfterInternal(
...
+
+    @Override
+    default String getNameColumn() {
+      return "fullyQualifiedName";
+    }
+
+    @RegisterRowMapper(QueryMapper.class)
+    @SqlQuery(
+        "SELECT query_entity.json, query_entity.fullyQualifiedName FROM query_entity INNER JOIN entity_relationship ON query_entity.id = entity_relationship.toId WHERE entity_relationship.fromId = :id and entity_relationship.relation = :relation and entity_relationship.toEntity = :entity and query_entity.fullyQualifiedName > :after order by query_entity.fullyQualifiedName LIMIT :limit")
+    List<QueryList> getAfterQueries(
```
Comment (mohityadav766) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java:1947]: please change the name

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java [hunk 4]
```diff
@@ -1899,6 +1925,47 @@ List<String> listPrefixesAfterInternal(
...
+        @Bind("id") String id,
+        @Bind("entity") String entity,
+        @Bind("after") String after,
+        @Bind("limit") int limit,
+        @Bind("relation") int relation);
+
+    @RegisterRowMapper(QueryMapper.class)
+    @SqlQuery(
+        "SELECT query_entity.json, query_entity.fullyQualifiedName FROM query_entity INNER JOIN entity_relationship ON query_entity.id = entity_relationship.toId WHERE entity_relationship.fromId = :id and entity_relationship.relation = :relation and entity_relationship.toEntity = :entity and query_entity.fullyQualifiedName < :before order by query_entity.fullyQualifiedName LIMIT :limit")
+    List<QueryList> getBeforeQueries(
```
Comment (mohityadav766) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java:1957]: change the function name

File: openmetadata-service/src/test/java/org/openmetadata/service/resources/query/QueryResourceTest.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  /**
+   * @param createdEntity
+   * @param request
+   * @param authHeaders
+   * @throws HttpResponseException
+   */
+  @Override
+  public void validateCreatedEntity(Query createdEntity, CreateQuery request, Map<String, String> authHeaders)
```
Comment (mohityadav766) [inline openmetadata-service/src/test/java/org/openmetadata/service/resources/query/QueryResourceTest.java:50]: can you please add the validation here

File: openmetadata-service/src/test/java/org/openmetadata/service/resources/query/QueryResourceTest.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+      throws HttpResponseException {}
+
+  /**
+   * @param expected
+   * @param updated
+   * @param authHeaders
+   * @throws HttpResponseException
+   */
+  @Override
+  public void compareEntities(Query expected, Query updated, Map<String, String> authHeaders)
```
Comment (mohityadav766) [inline openmetadata-service/src/test/java/org/openmetadata/service/resources/query/QueryResourceTest.java:60]: add here as well , please refer other test cases

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+import org.openmetadata.service.Entity;
+import org.openmetadata.service.resources.query.QueryResource;
+import org.openmetadata.service.util.EntityUtil;
+import org.openmetadata.service.util.FullyQualifiedName;
+import org.openmetadata.service.util.RestUtil;
+import org.openmetadata.service.util.ResultList;
+
+public class QueryRepository extends EntityRepository<Query> {
+
+  private static final String QUERY_PATCH_FIELDS = "";
```
Comment (mohityadav766) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java:20]: I think we need to add "vote" here

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+      queryList.remove(limit);
+      afterCursor = queryList.get(limit - 1).getFqn();
+    }
+    List<Query> queries = new ArrayList<>();
+    for (CollectionDAO.QueryList queryRow : queryList) {
+      queries.add(queryRow.getQuery());
+    }
+    return new ResultList<>(queries, beforeCursor, afterCursor, total);
+  }
+}
```
Comment (mohityadav766) [inline openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java:122]: please check this here , we can have some fields related to query that can be updated, if that's the case we need to implement EntityUpdater

File: openmetadata-spec/src/main/resources/json/schema/entity/data/query.json (hunk 2 above)
Comment (07Himank) [reply openmetadata-spec/src/main/resources/json/schema/entity/data/query.json]: done

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java (hunk 7 above)
Comment (07Himank) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java]: done

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java (hunk 8 above)
Comment (07Himank) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java:178]: done

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java (hunk 3 above)
Comment (07Himank) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java]: done

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    }
+    afterCursor = tableQueries.get(tableQueries.size() - 1).getFqn();
+    List<Query> queries = new ArrayList<>();
+    for (CollectionDAO.QueryList queryRow : tableQueries) {
+      queries.add(queryRow.getQuery());
+    }
+    return new ResultList<>(queries, beforeCursor, afterCursor, total);
+  }
+
+  private ResultList<Query> getAfterQueriesList(
```
[the remark this answers is not in the document]
Comment (07Himank) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/QueryRepository.java]: done

File: openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java (hunk 4 above)
Comment (07Himank) [reply openmetadata-service/src/main/java/org/openmetadata/service/jdbi3/CollectionDAO.java]: done

**Comment by 07Himank** [comment]
> Thanks @07Himank for the PR. We should have migration script that can load existing queries data into the new table and delete the old data.

**Comment by mohityadav766** [comment]
duplicated by #10449 ..closing

**Pull request closed without merging**